### PR TITLE
Quantization specific registration for CMSIS-NN

### PR DESCRIPTION
**Pull request by mansnils**
Adds three specific registrations for softmax kernel:
- Pure int8.
- Pure int16.
- And one with int8 input and int16 output.

In order to avoid duplicating code CalculateSoftmaxParams is made
public.

This progress towards https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/docs/rfc/002_16x8_quantization_port.md

BUG=add kernel variant to have the option to not increase binary size

### CODE REVIEW

File: tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc [hunk 1]
```diff
@@ -28,58 +28,104 @@ limitations under the License.
...
+          op_data.softmax_params, tflite::micro::GetTensorShape(input),
           tflite::micro::GetTensorData<float>(input),
           tflite::micro::GetTensorShape(output),
           tflite::micro::GetTensorData<float>(output));
       return kTfLiteOk;
     }
-    case kTfLiteInt8:
+    case kTfLiteInt8: {
+      if (output->type == kTfLiteInt8) {
+        arm_softmax_s8(tflite::micro::GetTensorData<int8_t>(input),
```
Comment (freddan80) [inline tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:101]: Would it be possible to call SoftmaxEvalInt8() here?

File: tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc [hunk 2]
```diff
@@ -28,58 +28,104 @@ limitations under the License.
...
+                       op_data.num_rows, op_data.row_size,
+                       op_data.softmax_params.input_multiplier,
+                       op_data.softmax_params.input_left_shift,
+                       op_data.softmax_params.diff_min,
+                       tflite::micro::GetTensorData<int8_t>(output));
+      } else {
+        arm_softmax_s8_s16(tflite::micro::GetTensorData<int8_t>(input),
```
Comment (freddan80) [inline tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:108]: dito for arm_softmax_s8_s16()?

File: tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc [hunk 3]
```diff
@@ -28,58 +28,104 @@ limitations under the License.
...
+      return kTfLiteOk;
+    }
     case kTfLiteInt16: {
-      SoftmaxQuantized(input, output, data);
+      const cmsis_nn_softmax_lut_s16 softmax_params = {
+          .exp_lut = op_data.softmax_params.exp_lut,
+          .one_by_one_lut = op_data.softmax_params.one_over_one_plus_x_lut};
+
+      TFLITE_DCHECK_EQ(
+          arm_softmax_s16(
```
Comment (freddan80) [inline tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:123]: dito for SoftmaxEvalInt16()

File: tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc (hunk 1 above)
Comment (mansnils) [reply tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:101]: I prefer not since it seems binary size will increase then.

File: tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc (hunk 2 above)
Comment (mansnils) [reply tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:108]: same comment here

File: tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc (hunk 3 above)
Comment (mansnils) [reply tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:123]: same

File: tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc [hunk 4]
```diff
@@ -89,17 +135,111 @@ TfLiteStatus SoftmaxEval(TfLiteContext* context, TfLiteNode* node) {
...
+TfLiteRegistration Register_SOFTMAX_INT16() {
+  softmax_registration.init = Init;
+  softmax_registration.free = nullptr;
+  softmax_registration.prepare = Prepare;
+  softmax_registration.invoke = SoftmaxEvalInt16;
+  softmax_registration.profiling_string = nullptr;
+  softmax_registration.builtin_code = 0;
+  softmax_registration.custom_name = nullptr;
+  softmax_registration.version = 0;
+  return softmax_registration;
```
Comment (paulinesho) [inline tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:242]: Is this change a style preference or is there a reason why we shouldn't use brace initialization here?

Comment (mansnils) [reply tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:242]: The reason is that the symbol softmax_registration is needed in this context for the linker to drop unused symbols. There is a TODO(b/169801227). So it is needed to get the code size benefits of this quantization specific registration.

Comment (paulinesho) [reply tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:242]: @mansnils Thanks for sharing that historical context. We're moving towards a common API for kernel registration in #1084. The good thing is that the new approach also reduces code size (see PR description for numbers) compared to both brace initialization and static initialization. I suggest we make the same change here for `softmax` if #1084 is merged before this.

Comment (mansnils) [reply tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc:242]: Well yes, I tried with and without the new common kernel registration API for cmsis-nn/fully_conneted.cc. Then I tried with and without the int8 specific registration in person detect example: https://github.com/tensorflow/tflite-micro/blob/10b1223343303b45e6f62a433e25efb5f87958d7/tensorflow/lite/micro/examples/person_detection/main_functions.cc#L78
And the same code size difference is seen. I can't see the internals of TODO(b/169801227) but I guess it can be closed then.

File: tensorflow/lite/micro/kernels/softmax_common.cc [hunk 5]
```diff
@@ -28,11 +28,48 @@ namespace {
...
 const int kInt16LUTArraySize = 513;
 
+}  // namespace
+
 TfLiteStatus CalculateSoftmaxParams(TfLiteContext* context,
                                     const TfLiteTensor* input,
                                     TfLiteTensor* output,
                                     const TfLiteSoftmaxParams* params,
                                     SoftmaxParams* op_data) {
+  // Only allocate LUTs for KTfLiteInt16 data type
```
Comment (deqiangc) [inline tensorflow/lite/micro/kernels/softmax_common.cc:38]: Could line 38 to line 71 be extracted as a separate function such as InitializeLutForInt16? This is a bit long and the function name will make it more readable.

Comment (mansnils) [reply tensorflow/lite/micro/kernels/softmax_common.cc:34]: Done

File: tensorflow/lite/micro/kernels/softmax_common.cc [hunk 6]
```diff
@@ -28,11 +28,58 @@ namespace {
 // Softmax parameter data that persists in user_data
 const int kInt16LUTArraySize = 513;
 
+__attribute__((always_inline)) inline TfLiteStatus InitializeLutForInt16(
```
[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/kernels/softmax_common.cc:31]: Ok but person detect binary increased by 296 bytes when removing the inline line for GCC. No difference for armclang.

Tested like this:
make -j$(nproc) -f tensorflow/lite/micro/tools/make/Makefile TOOLCHAIN=gcc OPTIMIZED_KERNEL_DIR=cmsis_nn TARGET=cortex_m_corstone_300 TARGET_ARCH=cortex-m55 person_detection_test && ll tensorflow/lite/micro/tools/make/gen/cortex_m_corstone_300_cortex-m55_default/bin/

Should we still remove it?

[the remark this answers is not in the document]
Comment (deqiangc) [reply tensorflow/lite/micro/kernels/softmax_common.cc:31]: Thanks for getting back Mans! I still prefer to remove it. In the mean time, we can create an issue to track this 296 bytes difference.

[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/kernels/softmax_common.cc]: Done

**Review by deqiangc — APPROVED** [review]
